allennlp/models/bidaf.py

```diff
@@ -0,0 +1,217 @@
+    attentions to put question information into the passage word representations (this is the only
+    part that is at all non-standard), pass this through another few layers of bi-LSTMs/GRUs, and
+    do a softmax over span start and span end.
+
+    Parameters
+    ----------
+    vocab : ``Vocabulary``
+        A Vocabulary, required in order to compute sizes for input/output projections.
+    text_field_embedder : ``TextFieldEmbedder``
+        Used to embed the ``question`` and ``passage`` ``TextFields`` we get as input to the model.
+    num_highway_layers : ``int``
+        The number of highway layers to use in between embedding the input and passing it through
+        the phrase layer.
+    phrase_layer : ``Seq2SeqEncoder``
+        The encoder (with its own internal stacking) that we will use in between embedding tokens
+        and doing the bidirectional attention.
+    attention_similarity_function : ``SimilarityFunction``
+        The similarity function that we will use when comparing encoded passage and question
+        representations.
+    modeling_layer : ``Seq2SeqEncoder``
+        The encoder (with its own internal stacking) that we will use in between the bidirectional
+        attention and predicting span start and end.
+    span_end_encoder : ``Seq2SeqEncoder``
+        The encoder that we will use to incorporate span start predictions into the passage state
+        before predicting span end.
+    """
+
+    def __init__(self, vocab: Vocabulary,
+                 text_field_embedder: TextFieldEmbedder,
+                 num_highway_layers: int,
+                 phrase_layer: Seq2SeqEncoder,
+                 attention_similarity_function: SimilarityFunction,
+                 modeling_layer: Seq2SeqEncoder,
+                 span_end_encoder: Seq2SeqEncoder) -> None:
+        super(BidirectionalAttentionFlow, self).__init__()
+
+        self._vocab = vocab
+        self._text_field_embedder = text_field_embedder
+        self._highway_layer = TimeDistributed(Highway(text_field_embedder.get_output_dim(),
+                                                      num_highway_layers))
+        self._phrase_layer = phrase_layer
+        self._matrix_attention = MatrixAttention(attention_similarity_function)
+        self._modeling_layer = modeling_layer
+        self._span_end_encoder = span_end_encoder
+
+        encoding_dim = phrase_layer.get_output_dim()
+        modeling_dim = modeling_layer.get_output_dim()
+        span_start_input_dim = encoding_dim * 4 + modeling_dim
+        self._span_start_predictor = TimeDistributed(torch.nn.Linear(span_start_input_dim, 1))
+
+        span_end_encoding_dim = span_end_encoder.get_output_dim()
+        span_end_input_dim = encoding_dim * 4 + span_end_encoding_dim
+        self._span_end_predictor = TimeDistributed(torch.nn.Linear(span_end_input_dim, 1))
+
+        # TODO(Mark): support masking once utility functions are merged.
```

Comment:
Copied directly from `simple_tagger.py` =).  You're the one who [knows what you meant](https://google.github.io/styleguide/pyguide.html?showone=TODO_Comments#TODO_Comments) there, so I kept your name.